open_spiel/games/othello.cc

```diff
@@ -0,0 +1,344 @@
+    case Direction::kUp:
+      return std::make_tuple(row - 1, col);
+    case Direction::kDown:
+      return std::make_tuple(row + 1, col);
+    case Direction::kLeft:
+      return std::make_tuple(row, col - 1);
+    case Direction::kRight:
+      return std::make_tuple(row, col + 1);
+    case Direction::kUpRight:
+      return std::make_tuple(row - 1, col + 1);
+    case Direction::kUpLeft:
+      return std::make_tuple(row - 1, col - 1);
+    case Direction::kDownRight:
+      return std::make_tuple(row + 1, col + 1);
+    case Direction::kDownLeft:
+      return std::make_tuple(row + 1, col - 1);
+    default:
+      SpielFatalError(absl::StrCat("Found unmatched case in GetNext."));
+  }
+}
+
+inline bool OthelloState::OnBoard(int row, int col) const {
+  return (0 <= row) && (row < kNumRows) && (0 <= col) && (col < kNumCols);
+}
+
+int OthelloState::CountSteps(Player player, int row, int col, Direction dir) const {
+  std::tuple<int, int> rowcol = GetNext(row, col, dir);
+  row = std::get<0>(rowcol);
+  col = std::get<1>(rowcol);
+
+  int count = 0;
+  CellState cell = PlayerToState(player);
+  while (OnBoard(row, col)) {
+    if (BoardAt(row, col) == cell) {
+      return count;
+    } else if (BoardAt(row, col) == CellState::kEmpty) {
+      return 0;
+    }
+      
+    count++;
+    rowcol = GetNext(row, col, dir);
+    row = std::get<0>(rowcol);
+    col = std::get<1>(rowcol);
+  }
+
+  return 0;
+}
+
+bool OthelloState::CanCapture(Player player, int move) const {
+  if (board_[move] != CellState::kEmpty) return false;
+
+  std::tuple<int, int> row_col = XYFromCode(move);
+  int row = std::get<0>(row_col);
+  int col = std::get<1>(row_col);
+
+  bool valid = false;
+  for (int direction = Direction::kUp; direction < Direction::kLast; direction++) {
+    if (CountSteps(player, row, col, static_cast<Direction>(direction)) != 0) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void OthelloState::Capture(Player player, int row, int col, Direction dir, int steps) {
+  std::tuple<int, int> rowcol = GetNext(row, col, dir);
+  row = std::get<0>(rowcol);
+  col = std::get<1>(rowcol);
+  CellState cell = PlayerToState(player);
+
+  for (int step = 0; step < steps; step++) {
+    if (BoardAt(row, col) == CellState::kEmpty || BoardAt(row, col) == cell) {
+      SpielFatalError(absl::StrCat("Cannot capture cell (", row, ", ", col, ")"));
+    }
+
+    board_[row * kNumCols + col] = cell;
+    
+    rowcol = GetNext(row, col, dir);
+    row = std::get<0>(rowcol);
+    col = std::get<1>(rowcol);
+  }
+}
+
+int OthelloState::DiskCount(Player player) const {
+  int count = 0;
+  CellState cell = PlayerToState(player);
+  for (int i = 0; i < kNumCells; i++) {
+    if (board_[i] == cell) count++;
+  }
+
+  return count;
+}
+
+bool OthelloState::NoValidActions() const {
+  return (LegalRegularActions(Player(0)).empty() & LegalRegularActions(Player(1)).empty());
+}
+
+std::tuple<int, int> OthelloState::XYFromCode(int move) const {
```

Comment:
Why a tuple instead of a pair?

You can also use the c++17 destructuring to get them back into x,y.

Alternatively, a specific class like https://github.com/deepmind/open_spiel/blob/master/open_spiel/games/havannah.h#L67 may be useful.

Reply:
I ended up discovering std::tie which allows you to do automatic unpacking from tuples/pairs. Thanks!

Reply:
Ah, I meant https://en.cppreference.com/w/cpp/language/structured_binding so you can even do:
auto [row, col] = GetNext(...)

though I think it'd be even better to add a Move class that decomposes it into row and col member variables. Then you'd do:
Move m(move);
f(BoardAt(m))
f(BoardAt(m.Next(direction)));
and pretty much never need to deal in row/col, but when you do you can do:
g(m.row, row.col)